plutus-core/plutus-core/src/PlutusCore/Core/Type.hs

```diff
@@ -219,9 +227,31 @@ mapFun f = go where
     go (Var ann name)             = Var ann name
     go (Constant ann con)         = Constant ann con
     go (Builtin ann fun)          = Builtin ann (f fun)
+    go (Constr ann ty i args)     = Constr ann ty i (map go args)
+    go (Case ann ty arg cs)       = Case ann ty (go arg) (map go cs)
 
 -- | This is a wrapper to mark the place where the binder is introduced (i.e. LamAbs/TyAbs)
 -- and not where it is actually used (TyVar/Var..).
 -- This marking allows us to skip the (de)serialization of binders at LamAbs/TyAbs positions
 -- iff 'name' is DeBruijn-encoded (level or index). See for example the instance of  'UntypedPlutusCore.Core.Instance.Flat'
 newtype Binder name = Binder { unBinder :: name }
+
+{- Note [Constr tag type]
+Constructor tags are not dynamically created, they can only come from the program itself.
```

Comment:
Since tags are always from the program, does it make sense to use Word16 or Word32 to save some space? Word32 is definitely sufficient, and Word16 should be, too - I can't imagine someone needing more than 65536 constructors.

Reply:
`Word64` is encoded like other integral types: using only as many bytes as are needed. So it won't save any space in the serialized program, and since it's a machine word I doubt we can actually save runtime memory by using a smaller type.

Reply:
I mentioned this in the comment.